src/main/java/org/cbioportal/cdd/service/internal/CDDServiceMskVocabularyImpl.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2018 - 2020 Memorial Sloan-Kettering Cancer Center.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF MERCHANTABILITY OR FITNESS
+ * FOR A PARTICULAR PURPOSE. The software and documentation provided hereunder
+ * is on an "as is" basis, and Memorial Sloan-Kettering Cancer Center has no
+ * obligations to provide maintenance, support, updates, enhancements or
+ * modifications. In no event shall Memorial Sloan-Kettering Cancer Center be
+ * liable to any party for direct, indirect, special, incidental or
+ * consequential damages, including lost profits, arising out of the use of this
+ * software and its documentation, even if Memorial Sloan-Kettering Cancer
+ * Center has been advised of the possibility of such damage.
+ */
+
+package org.cbioportal.cdd.service.internal;
+
+import java.util.*;
+import org.cbioportal.cdd.model.CancerStudy;
+import org.cbioportal.cdd.model.ClinicalAttributeMetadata;
+import org.cbioportal.cdd.model.MskVocabularyField;
+import org.cbioportal.cdd.repository.topbraid.MskVocabularyRepository;
+import org.cbioportal.cdd.service.ClinicalDataDictionaryService;
+import org.cbioportal.cdd.service.exception.CancerStudyNotFoundException;
+import org.cbioportal.cdd.service.exception.ClinicalAttributeNotFoundException;
+import org.cbioportal.cdd.service.exception.ClinicalMetadataSourceUnresponsiveException;
+import org.cbioportal.cdd.service.exception.FailedCacheRefreshException;
+import org.cbioportal.cdd.service.util.MSKVocabStudyUtil;
+import org.cbioportal.cdd.service.util.MskVocabularyConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+@Service
+@Qualifier("mskvocabulary")
+public class CDDServiceMskVocabularyImpl implements ClinicalDataDictionaryService {
+
+    @Autowired
+    private MSKVocabStudyUtil mskVocabStudyUtil;
+
+    @Autowired
+    private MskVocabularyConverter mskVocabularyConverter;
+
+    @Autowired
+    private MskVocabularyRepository mskVocabularyRepository;
+
+    private static final Logger logger = LoggerFactory.getLogger(CDDServiceMskVocabularyImpl.class);
+
+    private Map<String, ClinicalAttributeMetadata> clinicalAttributeMetadataCache = new HashMap<String, ClinicalAttributeMetadata>();
+
+    private void fillClinicalAttributeMetadataCache() {
+        List<MskVocabularyField> mskVocabularyList = mskVocabularyRepository.getClinicalAttributeMetadata();
+        if (mskVocabularyList.size() < clinicalAttributeMetadataCache.keySet().size() || mskVocabularyList == null) {
```

Comment:
only replace cache if the new set of attributes is larger in size than existing cache. We do not want to be losing attributes.

Reply:
Note : as is, we will need to refresh the caches via our api in order to reflect the deletion of an attribute in the upstream system (if there have not been more additions than deletions at the same time)

Reply:
This is existing behavior, think it's fine